src/Connect/command.cpp

```diff
@@ -0,0 +1,67 @@
+#include "command.hpp"
+
+#include <search_json.h>
+
+using json::Event;
+using json::Type;
+using std::string_view;
+
+namespace con {
+
+namespace {
+
+    // We are not completely sure what a token is in the notion of jsmn, besides we
+    // may need a bit more because it's allowed to put more data in there. It's on
+    // stack, so likely fine to overshoot a bit.
+    const constexpr size_t MAX_TOKENS = 60;
+
+}
+
+Command Command::gcode_command(CommandId id, const string_view &body) {
+    // TODO: We need to stuff the body somewhere. And we need to have an owned
+    // variant, since the command wil live longer than the caller. Some kind of
+    // std::string? But that's dynamic allocation :-(.
+    return Command {
+        id,
+        CommandType::Gcode,
+    };
+}
+
+Command Command::parse_json_command(CommandId id, const string_view &body) {
+    jsmntok_t tokens[MAX_TOKENS];
+
+    int parse_result;
+
+    {
+        jsmn_parser parser;
+        jsmn_init(&parser);
+
+        parse_result = jsmn_parse(&parser, body.data(), body.size(), tokens, sizeof tokens / sizeof *tokens);
+    } // Free the parser
+
+    CommandType command_type = CommandType::Unknown;
+
+    // Error from jsmn_parse will lead to -1 -> converted to 0, refused by json::search as Broken.
+    const bool success = json::search(body.data(), tokens, std::max(parse_result, 0), [&](const Event &event) {
+        if (event.depth != 1 || event.type != Type::String) {
+            return;
```

Comment:
Is this legal?
From C++11 draft:
6.6.3 The return statement [stmt.return]
1 A function returns to its caller by the return statement.
2 A return statement with neither an expression nor a braced-init-list can be used only in functions that do not
return a value, that is, a function with the return type cv void, a constructor (12.1), or a destructor (12.4).

Reply:
Yes, this is legal. This goes under the second point, function returning `void`. Well, it's lambda returning `void`, but that's the same kind of thing. Don't get yourself confused by the `json::search` itself (which returns `bool`), the lambda it takes as a parameter has no `-> ReturnType` specifier, which makes it return `void`.

If passing functions/lambdas into other functions isn't a commonly used technique in our team, I guess we can put it into the language window in the next meeting.